Aligning switch semantics with aria role

myasonik opened the pull request:
### Summary

Kibana was [seeing issues with switches](https://github.com/elastic/kibana/issues/41522). Though we weren't able to figure out the root cause of the issue, switches have a unique aria role that they weren't taking advantage of. So, instead of continuing to dig, I thought I'd update the semantics skirting the issue a bit.

Following [Inclusive Component's recommendation](https://inclusive-components.design/toggle-button/) I've moved switches away from `<input type=checkbox />` to `<button role=switch aria-checked=true|false>`.

- [x] Made sure the deprecation flow of shimming the event object works in Chrome/FF/Safari/IE 11/ChromiumEdge 

### Checklist

- [x] Checked in **dark mode**
- [x] Checked in **mobile**
- [x] Checked in **IE11** and **Firefox**
~- [ ] Props have proper **autodocs**~
~- [ ] Added **documentation** examples~
- [x] Added or updated **jest tests**
- [x] Checked for **breaking changes** and labeled appropriately
- [x] Checked for **accessibility** including keyboard-only and screenreader modes
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

myasonik commented on src/components/form/switch/switch.js:
```diff
@@ -15,6 +15,11 @@ export class EuiSwitch extends Component {
   }
 
+  onClick = e => {
+    e.target.checked = !this.props.checked;
```
So... This is kind of hacky but it prevents this from being a breaking change. Worth it?

cchaos replied:
@chandlerprall ? Seems wrong to me to set something in an on change handler based on props?

chandlerprall replied:
I think we should go through with a breaking change here - usage of `EuiSwitch` should be straight forward to find and update.

As for how the new/altered callback should look I'm torn between

`onClick={this.props.onChange}` // still call onChange, click event is passed through
`onClick={() => this.props.onChange}` // still call onChange, no argument passed
`onClick={() => this.props.onToggle}` // rename onChange -> onToggle, no event

myasonik replied:
I like the idea of changing the method name to `onToggle` because I think it's both a more accurate name and gives an easy distinction of which API you're using. Also, not sure if this is done, but could also for a deprecation window where both APIs are supported in parallel.

Thoughts on `onClick={e => this.props.onToggle(!this.props.checked, e)}`? That way, both the next state is passed through, which is likely what they need, but also the event is passed back because as an implementing developer, I always hate it when it's missing for my random, one-off use case.

chandlerprall replied:
We could support a deprecation window for this; There's ~75 uses of `EuiSwitch` in Kibana, and not all of them are as trivial as updating an inline event handler - deprecation would let us update these in smaller batches. We'll need to make sure how we add the `checked` value to the button target works in all browsers.

I'm fine with `onToggle` passing the updated checked value and also the source event object.

myasonik commented on src/global_styling/reset/_reset.scss:
```diff
@@ -71,6 +71,11 @@ body {
   outline: none;
+
+  // sass-lint:disable no-vendor-prefixes
+  &::-moz-focus-inner {
```
Firefox doesn't use `outline` for focus states; this removes the border firefox gives all focused elements

cchaos commented on src/components/form/switch/_switch.scss:
```diff
@@ -1,33 +1,20 @@
 .euiSwitch {
+  $euiColorSwitchIsOff: lightOrDarkTheme(transparentize($euiColorMediumShade, .8), transparentize($euiColorMediumShade, .3));
```
We are housing these switch variables the `_variables.scss` within the `form/` component's folder. I'd move this one there as well and rename it to align with the other names.

```suggestion
  $euiSwitchOffColor: lightOrDarkTheme(transparentize($euiColorMediumShade, .8), transparentize($euiColorMediumShade, .3));
```

cchaos commented on src/components/form/switch/_switch.scss:
```diff
@@ -80,50 +67,46 @@
-    .euiSwitch__thumb {
-      transform: scale(.95);
-    }
+  &:hover .euiSwitch__switch:active .euiSwitch__thumb {
```
Can you combine this selector with the `&:hover` one above so we can see them grouped?

cchaos commented on src/components/form/switch/switch.js:
```diff
@@ -40,35 +45,31 @@ export class EuiSwitch extends Component {
-          type="checkbox"
-          checked={checked}
+          aria-checked={checked}
+          className="euiSwitch__switch"
```
I'd rename this to `euiSwitch__button` so it's clear in the CSS what type of element it targets

myasonik commented:
@thompsongl Hmmm, it seems pretty smooth on my machine...

myasonik commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
 - Added `partial` glyph to `EuiIcon` ([#2152](https://github.com/elastic/eui/pull/2152))
 
+**Bug fixes**
```
Not sure I did correct still...

cchaos commented:
@myasonik I think you might have forgotten to push your latest changes?

cchaos commented on CHANGELOG.md:
```diff
@@ -1,6 +1,8 @@
 ## [`master`](https://github.com/elastic/eui/tree/master)
 
 - Added `partial` glyph to `EuiIcon` ([#2152](https://github.com/elastic/eui/pull/2152))
+- Fixed `EuiSwitch` semantics to align with aria roles ([#2193](https://github.com/elastic/eui/pull/2193))
```
This line will need to go under a `**Breaking changes**` header

cchaos commented on src-docs/src/views/form_controls/switch.js:
```diff
@@ -11,9 +11,9 @@ export default class extends Component {
   }
 
-  onChange = e => {
+  onToggle = newState => {
```
There are **a lot** more usages of EuiSwitch that will need to be updated to this new handler. Here's just a quick screenshot of a search within EUI:
<img width="369" alt="Screen Shot 2019-08-05 at 10 24 31 AM" src="https://user-images.githubusercontent.com/549577/62471894-9fafe800-b76b-11e9-8eea-2082b7605d08.png">

cchaos commented on src/components/form/switch/switch.js:
```diff
@@ -40,35 +40,31 @@ export class EuiSwitch extends Component {
+          className="euiSwitch__button"
+          role="switch"
           disabled={disabled}
-          onChange={onChange}
```
You'd need to swap `onChange` for `onToggle` from the props list both here
https://github.com/elastic/eui/blob/c03399529c56fd6e966885ab64eba2aa4b416821/src/components/form/switch/switch.js#L26

and here
https://github.com/elastic/eui/blob/c03399529c56fd6e966885ab64eba2aa4b416821/src/components/form/switch/switch.js#L78

And the TS defs need to be updated here:
https://github.com/elastic/eui/blob/c03399529c56fd6e966885ab64eba2aa4b416821/src/components/form/switch/index.d.ts

myasonik replied:
@chandlerprall What do you think of this direction before I make this change?

timroes commented:
Is there a specific reason (I didn't find any comment about that) why `onChange` has been renamed to `onToggle`? This feels very unrelated to fixing the semantics of the underlying HTML element and also has the problem (despite deprecation), that we need to fix A LOT OF code in Kibana. Also having every control using the same change callback name had the nice side effect that you knew it was `onChange` and you don't need to dig into documentation, to figure out if it's `onChange`, `onToggle`, `onChecked`, etc. depending on the concrete type?

myasonik commented:
@timroes It was a bit talked about [here](https://github.com/elastic/eui/pull/2193#discussion_r310487761). 

Basically:

1. Method name sounded more accurate
2. Having different method names lets you at a glance know if it's the deprecated API or the new one
3. Seems to give the cleanest deprecation flow (if using the old method name, shim the event handler to give the expected `checked` attribute; else, use the new method signature)

I don't think anyone held any strong feelings on this though. So, if you think keeping the same method name is better, I can revert the name change. Would you shim the event all the time then? Or not do it at all?

timroes commented:
Good question, I am also not strongly against renaming it, just wondering if we shouldn't then do a similar renaming for the other controls as well, so they are not using onChange (unless they are really input fields)? So if we would keep `onChange` I assume we should also keep the "shimmed" event. But I don't have too strong feeling in either direction, so I'll totally leave the decision to you.

myasonik commented:
@chandlerprall got any strong feelings?

I'm leaning towards leaving it how I currently have it (deprecated `onChange` with shimmed `event` object, new `onToggle` method). 

Not sure what to do about other components though. Nothing right now but, in the future... 🤷‍♂

chandlerprall commented:
It definitely is nice to keep the form controls' API the same and expose `onChange`, but in this case I think that becomes disingenuous as the `onChange` for switch no longer matches the other form controls unless we always shim the `checked` prop onto the button, which doesn't seem like the right long-term choice.

Which brings up another point - the shimming of a `checked` onto the button won't enable backwards compatibility in TypeScript.

This PR updates `EuiSwitchProps` to `ButtonHTMLAttributes<HTMLInputElement> & {` but it should be `ButtonHTMLAttributes<HTMLButtonElement> & {`. With that additional change, accessing the change event's target.checked in a typescript file fails

> Error:(10, 49) TS2339: Property 'checked' does not exist on type 'EventTarget & HTMLButtonElement'.

We would need to typedef the shim with

```
	  export type EuiSwitchProps = CommonProps &
	    Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'onChange'> & {
	      label?: ReactNode;
	      checked: boolean;
	      onChange?: (event: React.FormEvent<HTMLButtonElement & { checked: boolean }>) => void;
	      onToggle?: Function;
	      disabled?: boolean;
	      compressed?: boolean;
	    };
```

chandlerprall commented:
Alright, we discussed this in the design meeting and we're all leaning toward maintaining the better developer experience, which is keeping this as `onChange` with no deprecation.

This assumes:

* All browsers are fine with setting `checked` to `true`/`false` on a button element
* TypeScript defs in EUI can communicate this additional value correctly (see my previous comment)
* React's synthetic event wrapping doesn't do anything weird to prevent this

myasonik commented:
OK, I've removed the breaking change and have left the shimed `checked` property.

I also changed the documentation to not actually use that property anymore though, instead relying on locally toggling the state, to quietly discourage its use. (Most implementations of swtich seem to do this anyway.)
